src/include/txn.i

```diff
@@ -343,51 +343,58 @@ __wt_txn_unmodify(WT_SESSION_IMPL *session)
 }
 
 /*
- * __wt_txn_op_commit_page_del --
- *	Make the transaction ID and timestamp updates necessary to a ref that
- *      was created by a fast delete truncate operation.
+ * __wt_txn_op_apply_prepare_state --
+ *	Apply the correct prepare state and the timestamp to the ref and to any
+ *      updates in the page del update list.
  */
 static inline void
-__wt_txn_op_commit_page_del(WT_SESSION_IMPL *session, WT_REF *ref)
+__wt_txn_op_apply_prepare_state(
+    WT_SESSION_IMPL *session, WT_REF *ref, bool commit)
 {
 	WT_TXN *txn;
 	WT_UPDATE **updp;
+	wt_timestamp_t ts;
 	uint32_t previous_state;
+	uint8_t prepare_state;
 
 	txn = &session->txn;
 
-	/* Avoid locking the page if a previous eviction already cleaned up. */
-	if (ref->page_del->update_list == NULL)
-		return;
-
 	/*
 	 * Lock the ref to ensure we don't race with eviction freeing the
-	 * page deleted update list.
+	 * page deleted update list or with a page instantiate.
 	 */
 	for (;; __wt_yield()) {
 		previous_state = ref->state;
+		WT_ASSERT(session, previous_state != WT_REF_READING);
 		if (previous_state != WT_REF_LOCKED &&
+		    (previous_state == WT_REF_MEM ||
+		    previous_state == WT_REF_DELETED) &&
```

Comment:
I don't think this is OK. I think the ref could be `WT_REF_LOOKASIDE` or `WT_REF_LIMBO`. Even if it couldn't the consequence of a ref being in that state unexpectedly is that this code would spin forever.

Reply:
I guess this needs a discussion. This check is from the notes I took when we had a meeting with @michaelcahill

Reply:
We talked about various options here but I thought we ended up agreeing to add the assertion above (`previous_state != WT_REF_READING`) but not make these changes.

Reply:
Fixed.